pkg/kubelet/kuberuntime/instrumented_services_test.go

```diff
@@ -24,24 +24,28 @@ import (
 
 	"github.com/stretchr/testify/assert"
 
-	"k8s.io/component-base/metrics/legacyregistry"
+	compbasemetrics "k8s.io/component-base/metrics"
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
 	"k8s.io/kubernetes/pkg/kubelet/metrics"
 )
 
 func TestRecordOperation(t *testing.T) {
-	legacyregistry.MustRegister(metrics.RuntimeOperations)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsDuration)
-	legacyregistry.MustRegister(metrics.RuntimeOperationsErrors)
+	// Use local registry
+	var registry = compbasemetrics.NewKubeRegistry()
+	var gather compbasemetrics.Gatherer = registry
+	defer registry.Reset()
```

Comment:
Both seem to be OK.
I prefer `defer it to the end` because we want to fix the test case multiple times.
The reset at the end of the test case will clear the metrics/env after running the case. 
This is also a very clear way.

Reply:
Doing the Reset at the end works if every test does a Reset at the end.  Doing a Reset at the start works regardless of what other tests do.  A local condition is better than a global one.

Reply:
Both ways are okay.  

I prefer `defer`. Because
1. Other test files willn't affect this file. Because RuntimeOperations is supposed to be tested in this file.
2. It requires tests in this file doing Reset at the end.  We have done it since registration appears only once.  Or, it could keep code style more consistent if registration appears many times.

Reply:
Both ways will work if, in a given process, no other function adds data to `metrics.RuntimeOperations`, `metrics.RuntimeOperationsDuration`, or `metrics.RuntimeOperationsErrors` before TestRecordOperation runs.  That is a global condition.  Note that `metrics.RuntimeOperations` is registered in legacyregistry in some other code invoked by another test.  Other test files _can_ affect this one, if the Reset is done at the end in this one.  Local conditions are better than global ones.

Reply:
> Both ways will work if, in a given process, no other function adds data to `metrics.RuntimeOperations`, `metrics.RuntimeOperationsDuration`, or `metrics.RuntimeOperationsErrors` before TestRecordOperation runs. That is a global condition. Note that `metrics.RuntimeOperations` is registered in legacyregistry in some other code invoked by another test. Other test files _can_ affect this one, if the Reset is done at the end in this one. Local conditions are better than global ones.

I find this argument persuasive.  Globals present numerous problems, but clearing the state before starting a test keeps the scope of control inside this Test instead of requiring every other test to "be perfect".

Reply:
> clearing the state before starting a test keeps the scope of control inside this Test instead of requiring every other test to "be perfect".

yeah, we can't rely that other tests are right, but we also need to clean up the state once the test ends, same as we do with the listeners, per example

```
defer l.Close()
```
I feel that we need both
```
diff --git a/pkg/kubelet/kuberuntime/instrumented_services_test.go b/pkg/kubelet/kuberuntime/instrumented_services_test.go
index e95d6bdb74a..1801f995d5f 100644
--- a/pkg/kubelet/kuberuntime/instrumented_services_test.go
+++ b/pkg/kubelet/kuberuntime/instrumented_services_test.go
@@ -37,6 +37,7 @@ func TestRecordOperation(t *testing.T) {
        registry.MustRegister(metrics.RuntimeOperations)
        registry.MustRegister(metrics.RuntimeOperationsDuration)
        registry.MustRegister(metrics.RuntimeOperationsErrors)
+ registry.Reset()
 
        l, err := net.Listen("tcp", "127.0.0.1:0")
```

Reply:
I do not follow the analogy to calling Close.  The Close method is about releasing expensive resources that an active connection holds.  Reset is not analogous, it does not release expensive resources.  I mean, there may be some internal side-effects of resetting some metrics, but it is nothing like open network connections that can accumulate and cause problems.

Reply:
You are right, my bad,  I misinterpreted the reset on metrics. 

Agree with you and David, the test has to clear the state before starting and not depend that other tests do the same after finishing